aten/src/ATen/cuda/Atomic.cuh

```diff
@@ -71,86 +71,88 @@ struct AtomicFPOp<double> {
+      assumed = old;                                                                                                   \
+      newval = static_cast<uint32_t>(OP(val, static_cast<T>(old)));                                                    \
+      old = atomicCAS(address_as_ui, assumed, newval);                                                                 \
+    } while (assumed != old);                                                                                          \
+  }                                                                                                                    \
+};                                                                                                                     \
+                                                                                                                       \
+template<typename T>                                                                                                   \
+struct Atomic##NAME##IntegerImpl<T, 8> {                                                                               \
+  inline __device__ void operator()(T *address, T val) {                                                               \
+    unsigned long long * address_as_ui = (unsigned long long *) (address);                                             \
+    unsigned long long old = *address_as_ui;                                                                           \
+    unsigned long long newval;                                                                                         \
+    unsigned long long assumed;                                                                                        \
+                                                                                                                       \
+    do {                                                                                                               \
+      assumed = old;                                                                                                   \
+      newval = static_cast<uint64_t>(OP(val, static_cast<T>(old)));                                                    \
+      old = atomicCAS(address_as_ui, assumed, newval);                                                                 \
+    } while (assumed != old);                                                                                          \
+  }                                                                                                                    \
 };
 
-template<typename T>
-struct AtomicAddIntegerImpl<T, 8> {
-  inline __device__ void operator()(T *address, T val) {
-    unsigned long long * address_as_ui = (unsigned long long *) (address);
-    unsigned long long old = *address_as_ui;
-    unsigned long long newval;
-    unsigned long long assumed;
+#define OP(X, Y) X + Y
```

Comment:
Why not pass "X + Y" as an argument to the `ATOMIC_INTEGER_IMPL` macro?

Reply:
You'd have to constraint yourself to using "X" and "Y" around OP (not ideal).

Otherwise you can also look at `AtomicFPOp` where the function is passed as a template parameter. Then 

```
AtomicMulIntegerImpl<uint8_t, sizeof(uint8_t)>()(address, val);
```
becomes
```
AtomicMulIntegerImpl<uint8_t, sizeof(uint8_t)>()(address, 
                                                 val,
                                                 [](uint8_t a, uint8_t b) {
                                                     return a * b;
                                                 });

```

Reply:
You can then use a macro again to instantiate all the dtypes